### PR TITLE
Testing and Integrating new qemu_thread_set_affinity definition

### PR DESCRIPTION
**Pull request by roja-zededa**
Xentools 4.19.0 which uses QEMU 8.0.4 has new implementation for CPU pinning. A big difference between our implementation and the QEMU vanilla version is that our implementation can pin vCPU-to-pCPU 1-to-1, while the QEMU implementation allows only a set of CPUs to be assigned to an entire VM (and the VM's vCPU threads can still migrate among the given set of pCPUs). Need to verify and intergrate if the new qemu_thread_set_affinity definition is suitable for our use case

### CODE REVIEW

**Comment by OhmSpectator** [comment]
The most critical use case for the new CPU pinning is realtime, as this implementation does not really pin CPUs. I would switch to it if it does not worsen the results of RT tests. @rouming can provide more information, how he performed the testing.
Also, if using the native qemu implementation, the qemu parameters should be used in native way, as well. You can find it in the PR I shared with you earlier.

**Comment by roja-zededa** [comment]
> The most critical use case for the new CPU pinning is realtime, as this implementation does not really pin CPUs. I would switch to it if it does not worsen the results of RT tests. @rouming can provide more information, how he performed the testing.
> 
> Also, if using the native qemu implementation, the qemu parameters should be used in native way, as well. You can find it in the PR I shared with you earlier: [rucoder#1](https://github.com/rucoder/eve/pull/1)

@rouming Could you please provide more info on RT CPU Pinning tests?

**Comment by rouming** [comment]
> > The most critical use case for the new CPU pinning is realtime, as this implementation does not really pin CPUs. I would switch to it if it does not worsen the results of RT tests. @rouming can provide more information, how he performed the testing.
> > Also, if using the native qemu implementation, the qemu parameters should be used in native way, as well. You can find it in the PR I shared with you earlier: [rucoder#1](https://github.com/rucoder/eve/pull/1)
> 
> @rouming Could you please provide more info on RT CPU Pinning tests?

The RT suite is here: https://git.kernel.org/pub/scm/utils/rt-tests/rt-tests.git/ , the standard de facto is cyclictest, you can search how to use it if you wish, but I can tell in advance, that pinning the whole process with all its thread to set of cpus comparing to 1to1 mapping has no visible impact on any test results, just because scheduler is smart enough and wont perform a task migration to another cpu if load on all vcpus is equal (usually this is the case). even if task migrates, that does not happen frequently, so impact is misarable.

**Comment by OhmSpectator** [comment]
It would be nice to remove the custom patches in this case.

But I don't see how the QEMU vanilla pinning is used here? To run with the native CPU pinning, QEMU should start with `-object thread-context,id=tc1,cpu-affinity=0-1,cpu-affinity=6-7` option or similar...

**Comment by rouming** [comment]
@roja-zededa I see this PR contains the commit message from another PR "Upgraded Xen-tools from 4.15 to 4.19.0", not sure what is the purpose of duplicating these if this is not a mistake.

**Comment by roja-zededa** [comment]
> @roja-zededa I see this PR contains the commit message from another PR "Upgraded Xen-tools from 4.15 to 4.19.0", not sure what is the purpose of duplicating these if this is not a mistake.

No, it wasn't a duplicate. Upgrading xentool version to 4.19.0 introduced a new vanilla qemu pinning feature from QEMU 8.0, we're testing here if we need to keep the vanilla pinning feature or go with Nikolay's 1-1 custom pinning patch which is already included in the previous xentools versions.

**Comment by rouming** [comment]
> > @roja-zededa I see this PR contains the commit message from another PR "Upgraded Xen-tools from 4.15 to 4.19.0", not sure what is the purpose of duplicating these if this is not a mistake.
> 
> @rouming No, it wasn't a duplicate. Upgrading xentool version to 4.19.0 introduced a new vanilla qemu pinning feature from QEMU 8.0, we're testing here if we need to keep the vanilla pinning feature or go with Nikolay's 1-1 custom pinning patch which is already included in the previous xentools versions.

Now I see you pushed a commit which corresponds to the PR description, nice.

Can you please share what are the results of testing and how exactly do you test two different implementations.

**Comment by OhmSpectator** [comment]
I would just cherry-pick the original commit and save the commit message.
```
git remote add ohmspectator git@github.com:OhmSpectator/eve.git
git fetch ohmspectator
git cherry-pick 32ec62b016016792d2c07512428f2e50559e7c34
```

File: pkg/pillar/hypervisor/xen.go [hunk 1]
```diff
@@ -260,8 +260,12 @@ func (ctx xenContext) CreateDomConfig(domainName string,
 		maxCpus = vCpus
 	}
 	file.WriteString(fmt.Sprintf("maxvcpus = %d\n", maxCpus))
-	if config.CPUs != "" {
-		file.WriteString(fmt.Sprintf("cpus = \"%s\"\n", config.CPUs))
+	if len(config.CPUs) > 0 {
+		cpusString := make([]string, 0)
+		for _, curCpu := range config.CPUs {
```
Comment (OhmSpectator) [inline pkg/pillar/hypervisor/xen.go:265]: Yetus asks to rename it to `curCPU` =\

File: pkg/pillar/hypervisor/xen.go [hunk 2]
```diff
@@ -260,8 +260,12 @@ func (ctx xenContext) CreateDomConfig(domainName string,
 		maxCpus = vCpus
 	}
 	file.WriteString(fmt.Sprintf("maxvcpus = %d\n", maxCpus))
-	if config.CPUs != "" {
-		file.WriteString(fmt.Sprintf("cpus = \"%s\"\n", config.CPUs))
+	if len(config.CPUs) > 0 {
+		cpusString := make([]string, 0)
+		for _, curCPU := range config.CPUs {
+			cpusString = append(cpusString, strconv.Itoa(curCpu))
```
Comment (OhmSpectator) [inline pkg/pillar/hypervisor/xen.go:266]: Here as well))

**Review by OhmSpectator — COMMENTED** [review]
In general, it looks fine to me!

But I would ask to create a copy of this document:
https://docs.google.com/document/d/1EAf1Oxyn6PWc64caqlXbO791tvOV2HiMRMUwwU1IM4E/edit?usp=sharing and update it accordingly. 

Perform the test flow described there, evaluate the results, check that they are as we expect them (all VCPU-related threads running only on the set of given CPUs), and update the document accordingly. 

The difference I would expect for example here:

```
VMs with CPU pinning, creation.
Deploy a VM with cpu-pinning enabled and 2 CPUs.
Check the output of the snippet:
======= QEMU 4428 threads: =======
pid 4428's current affinity list: 0,1,4-15
pid 4442's current affinity list: 0,1,4-15
pid 4453's current affinity list: 0,1,4-15
pid 4454's current affinity list: 0,1,4-15
pid 4455's current affinity list: 0,1,4-15
pid 4457's current affinity list: 0,1,4-15
======= QEMU 15476 threads: =======
pid 15476's current affinity list: 2,3
pid 15490's current affinity list: 2,3
pid 15501's current affinity list: 2,3
pid 15502's current affinity list: 2 <<<<------ should be 2,3 now
pid 15503's current affinity list: 3 <<<<------ should be 2,3 now
pid 15505's current affinity list: 2,3

```
The description under the test should be also fixed ("2 threads of the VM should have affinity lists containing only 1 CPU, and the rest of the threads: both of them simultaneously." is not true anymore).

**Comment by roja-zededa** [comment]
Trying to launch a VM but got this error: 

`{"source":"369dc470-4c17-4278-b863-0af36e746804.1.1","content":"qemu-system-x86_64: Property 'pc-q35-3.1-machine.cpumask' not found","msgid":2,"timestamp":{"seconds":1727980005,"nanos":666486415}}`

**Comment by OhmSpectator** [comment]
> Trying to launch a VM but got this error:
> 
> `{"source":"369dc470-4c17-4278-b863-0af36e746804.1.1","content":"qemu-system-x86_64: -no-hpet: warning: -no-hpet is deprecated, use '-machine hpet=off' instead","msgid":1,"timestamp":{"seconds":1727980005,"nanos":666473281}}`
> 
> `{"source":"369dc470-4c17-4278-b863-0af36e746804.1.1","content":"qemu-system-x86_64: Property 'pc-q35-3.1-machine.cpumask' not found","msgid":2,"timestamp":{"seconds":1727980005,"nanos":666486415}}`

Maybe something else has changed between QEMU 9.0.4, where I tested it, and the version introduced with the Xen-tool upgrade. I hope you can figure out the reason =)

Also, I want to mention a new test suite introduced by Vignesh. It automated, among others, the CPU pinning affinity checks. We should have a demo soon on how to use/adopt it. Now it's available from the Zededa Jenkins UI with the title: "Ztaf"

**Comment by OhmSpectator** [comment]
@roja-zededa, any progress on manual testing?

**Comment by roja-zededa** [comment]
` qemu_thread_set_affinity(cpu->thread, NULL, cpu->cpumask);` definition has been changed in the newer version. As of now, passing NULL for CPUlist but that might cause some issues. Need to change patch 15 to accommodate the new definition.

**Comment by OhmSpectator** [comment]
> ` qemu_thread_set_affinity(cpu->thread, NULL, cpu->cpumask);` definition has been changed in the newer version. As of now, passing NULL for CPUlist but that might cause some issues. Need to change patch 15 to accommodate the new definition.

I would make it as a new patch without changing patch 15.

**Comment by roja-zededa** [comment]
@OhmSpectator The board I have right now doesn't have 10 cores. Would it be okay to manually perform the test on a four-core DellOptiplex? If that doesn't work we can directly kickoff the new cpupinning test desgined by vignesh.

**Comment by OhmSpectator** [comment]
You can also run EVE in QEMU with a given amount of VCPUs =)

**Comment by rene** [comment]
@roja-zededa , in the commit title "[pillar: Adopt pillar to use native cpupinning](https://github.com/lf-edge/eve/pull/4212/commits/90db3a39f78e57f5f1a32a9b52d6a8cf6d7f5f23)", I guess you meant: "pillar: **Adapt** pillar to use..."

File: pkg/pillar/types/domainmgrtypes.go [hunk 3]
```diff
@@ -242,7 +242,7 @@ type VmConfig struct {
 	ExtraArgs  string // added to bootargs
 	BootLoader string // default ""
 	// For CPU pinning
-	CPUs string // default "", list of "1,2"
+	CPUs []int // default nil, list of [1,2]
```
Comment (rene) [inline pkg/pillar/types/domainmgrtypes.go:245]: @roja-zededa , this change implies on change the API as well: https://github.com/lf-edge/eve-api/blob/main/proto/config/vm.proto#L48

I'm not sure how this impacts the controller.... I wouldn't change the type, you can still keep as string and parse it accordingly....

Comment (uncleDecart) [reply pkg/pillar/types/domainmgrtypes.go:245]: +1 changes on API would require changes on controller

**Review by uncleDecart — COMMENTED** [review]
I would also add test to domainmgr checking that CPUs were set as expected in VmConfig.CPUs (something like [this](https://github.com/lf-edge/eve/blob/8e7e50049f90093127e66932792325ab47dc62d3/pkg/pillar/cmd/domainmgr/domainmgr_test.go#L593)) and that generated hypervisor configs are as expected

File: pkg/pillar/types/domainmgrtypes.go (hunk 3 above)
Comment (roja-zededa) [reply pkg/pillar/types/domainmgrtypes.go:245]: Nikolay authored this commit.  @OhmSpectator  Could you please take a look at it?

Comment (eriknordmark) [reply pkg/pillar/types/domainmgrtypes.go:245]: cmd/zedagent/parseconfig.go parses the received protobuf and produces the golang structs such as this. Thus it can be made to parse the existing protbuf string format and produce an array of integers.
Thus no need to change the EVE API.

Comment (OhmSpectator) [reply pkg/pillar/types/domainmgrtypes.go:245]: This field in the protobuf that @rene pointed to is not filled out by the controller at the moment. So we don't even have to change the parsing logic. It's currently used only internally in the DomainStatus.

**Comment by OhmSpectator** [comment]
> ======= QEMU 1669 threads: =======
> pid 1669's current affinity list: 2
> pid 1682's current affinity list: 2
> pid 1683's current affinity list: 2
> pid 1688's current affinity list: 2
> pid 1689's current affinity list: 2
> pid 2258's current affinity list: 0-3
> ======= QEMU 32076 threads: =======
> pid 32076's current affinity list: 0,1
> pid 32094's current affinity list: 0,1
> pid 32098's current affinity list: 0,1
> pid 32099's current affinity list: 0,1
> pid 2881's current affinity list: 0,1

This looks strange as I would expect 
```
> ======= QEMU 1669 threads: =======
> pid 1669's current affinity list: 2
> pid 1682's current affinity list: 2
> pid 1683's current affinity list: 2
> pid 1688's current affinity list: 2
> pid 1689's current affinity list: 2
> pid 2258's current affinity list: 0-3
> ======= QEMU 32076 threads: =======
> pid 32076's current affinity list: 0,1,3
> pid 32094's current affinity list: 0,1,3
> pid 32098's current affinity list: 0,1,3
> pid 32099's current affinity list: 0,1,3
> pid 2881's current affinity list: 0,1,3
```
But maybe the CPUs were redistributed a little bit later.

**Review by OhmSpectator — COMMENTED** [review]
@roja-zededa , thanks for the tests. The results look fine. Nevertheless, I would prefer to run the tests as they are described in the doc, as they also involve some corner cases. 

Also, could you please run this snippet to check the affinities (should be run from the debug container with the `procps` package installed):
```
for pid in $(pgrep qemu); do 
    echo "======= QEMU $pid threads: ======="; 
    for spid in $(ps -T -o spid= -p "$pid"); do 
        echo -n "Thread $spid affinity: "; taskset -pc "$spid"; 
        cgpath=$(grep -m1 cpuset /proc/$spid/cgroup | cut -d: -f3)
        echo "Cgroup cpuset: $(cat /sys/fs/cgroup/cpuset${cgpath}/cpuset.cpus)"
    done; 
done
```
I want to see how it will look like for the non-VCPU threads, as they are now not affected by the CPU pinning:
```
======= QEMU 1669 threads: =======
pid 1669's current affinity list: 2
pid 1682's current affinity list: 2
pid 1683's current affinity list: 2
pid 1688's current affinity list: 2
pid 1689's current affinity list: 2
pid 2258's current affinity list: 0-3 <---- here
pid 4727's current affinity list: 2
pid 4728's current affinity list: 2
pid 4729's current affinity list: 2
pid 4762's current affinity list: 0-3 <---- here
```
I also want to understand, why most of the threads have affinity corresponding to the CPU mask and why not others... I expected VCPU thread to affected only, but we definetly see more. If you could, would be nice if you check how the new version of QEMU handles this option:
```
-object thread-context,id=tc1,...
```
may be we should do the same for `tc0`, or only for `tc0`?

**Comment by roja-zededa** [comment]
> > ======= QEMU 1669 threads: =======
> > pid 1669's current affinity list: 2
> > pid 1682's current affinity list: 2
> > pid 1683's current affinity list: 2
> > pid 1688's current affinity list: 2
> > pid 1689's current affinity list: 2
> > pid 2258's current affinity list: 0-3
> > ======= QEMU 32076 threads: =======
> > pid 32076's current affinity list: 0,1
> > pid 32094's current affinity list: 0,1
> > pid 32098's current affinity list: 0,1
> > pid 32099's current affinity list: 0,1
> > pid 2881's current affinity list: 0,1
> 
> This looks strange as I would expect
> 
> ```
> > ======= QEMU 1669 threads: =======
> > pid 1669's current affinity list: 2
> > pid 1682's current affinity list: 2
> > pid 1683's current affinity list: 2
> > pid 1688's current affinity list: 2
> > pid 1689's current affinity list: 2
> > pid 2258's current affinity list: 0-3
> > ======= QEMU 32076 threads: =======
> > pid 32076's current affinity list: 0,1,3
> > pid 32094's current affinity list: 0,1,3
> > pid 32098's current affinity list: 0,1,3
> > pid 32099's current affinity list: 0,1,3
> > pid 2881's current affinity list: 0,1,3
> ```
> 
> But maybe the CPUs were redistributed a little bit later.

The deleted cpus are being reclaimed. Here is the trace. 
QEMU 22745 - No CPU pin 
QEMU 23728 - 1 CPU pinned
QEMU 24930  - 1CPU Pinned

======= QEMU 22745 threads: =======
pid 22745's current affinity list: 0,3
pid 22774's current affinity list: 0,3
pid 22778's current affinity list: 0,3
pid 22779's current affinity list: 0,3
pid 24456's current affinity list: 0,3
pid 25204's current affinity list: 0,3
pid 25216's current affinity list: 0-3
======= QEMU 23728 threads: =======
pid 23728's current affinity list: 1
pid 23742's current affinity list: 1
pid 23743's current affinity list: 1
pid 23747's current affinity list: 1
pid 23748's current affinity list: 1
pid 24056's current affinity list: 0-3
======= QEMU 24930 threads: =======
pid 24930's current affinity list: 2
pid 24944's current affinity list: 2
pid 24945's current affinity list: 2
pid 24949's current affinity list: 2
pid 24950's current affinity list: 2
pid 24952's current affinity list: 2
pid 25024's current affinity list: 0-3
pid 25025's current affinity list: 2
pid 25026's current affinity list: 2
pid 25027's current affinity list: 2


Removal:

Deletea VM

======= QEMU 22745 threads: =======
pid 22745's current affinity list: 0,1,3
pid 22774's current affinity list: 0,1,3
pid 22778's current affinity list: 0,1,3
pid 22779's current affinity list: 0,1,3
pid 27625's current affinity list: 0-3
======= QEMU 24930 threads: =======
pid 24930's current affinity list: 2
pid 24944's current affinity list: 2
pid 24945's current affinity list: 2
pid 24949's current affinity list: 2
pid 24950's current affinity list: 2
pid 25420's current affinity list: 0-3
pid 27674's current affinity list: 2

**Comment by roja-zededa** [comment]
@OhmSpectator Covered all the test cases in this document. The results look good to me. https://docs.google.com/document/d/1qoCq7SPk6UR2LCb6APVbmLLgo7W7wONJUncTrXCbPeM/edit?tab=t.0#heading=h.p1hte28jbxjx

**Comment by OhmSpectator** [comment]
@roja-zededa, please address the rest of my comments here:
https://github.com/lf-edge/eve/pull/4212#pullrequestreview-2404369311

1. I want to understand what are the the threads that run only on a given subset of CPUs and what are the threads that run on all CPUs. To understand it, please check how the thread contexts work. We use `tc1`, but we should understand which exactly threads are affected by it.

2. Check not only the CPU affinities but the CPUset in cgroup. I have already provided a snippet for that: 
```
for pid in $(pgrep qemu); do 
    echo "======= QEMU $pid threads: ======="; 
    for spid in $(ps -T -o spid= -p "$pid"); do 
        echo -n "Thread $spid affinity: "; taskset -pc "$spid"; 
        cgpath=$(grep -m1 cpuset /proc/$spid/cgroup | cut -d: -f3)
        echo "Cgroup cpuset: $(cat /sys/fs/cgroup/cpuset${cgpath}/cpuset.cpus)"
    done; 
done
```

**Comment by roja-zededa** [comment]
> @roja-zededa, please address the rest of my comments here: [#4212 (review)](https://github.com/lf-edge/eve/pull/4212#pullrequestreview-2404369311)
> 
> 1. I want to understand what are the the threads that run only on a given subset of CPUs and what are the threads that run on all CPUs. To understand it, please check how the thread contexts work. We use `tc1`, but we should understand which exactly threads are affected by it.
> 2. Check not only the CPU affinities but the CPUset in cgroup. I have already provided a snippet for that:
> 
> ```
> for pid in $(pgrep qemu); do 
>     echo "======= QEMU $pid threads: ======="; 
>     for spid in $(ps -T -o spid= -p "$pid"); do 
>         echo -n "Thread $spid affinity: "; taskset -pc "$spid"; 
>         cgpath=$(grep -m1 cpuset /proc/$spid/cgroup | cut -d: -f3)
>         echo "Cgroup cpuset: $(cat /sys/fs/cgroup/cpuset${cgpath}/cpuset.cpus)"
>     done; 
> done
> ```

@OhmSpectator Updated this doc:https://docs.google.com/document/d/1qoCq7SPk6UR2LCb6APVbmLLgo7W7wONJUncTrXCbPeM/edit?tab=t.0#heading=h.p1hte28jbxjx  with the CPUset group info and CPU Affinity. Please refer to it. I don't see any issue with it so far.

**Comment by roja-zededa** [comment]
> > @roja-zededa, please address the rest of my comments here: [#4212 (review)](https://github.com/lf-edge/eve/pull/4212#pullrequestreview-2404369311)
> > 
> > 1. I want to understand what are the the threads that run only on a given subset of CPUs and what are the threads that run on all CPUs. To understand it, please check how the thread contexts work. We use `tc1`, but we should understand which exactly threads are affected by it.
> > 2. Check not only the CPU affinities but the CPUset in cgroup. I have already provided a snippet for that:
> > 
> > ```
> > for pid in $(pgrep qemu); do 
> >     echo "======= QEMU $pid threads: ======="; 
> >     for spid in $(ps -T -o spid= -p "$pid"); do 
> >         echo -n "Thread $spid affinity: "; taskset -pc "$spid"; 
> >         cgpath=$(grep -m1 cpuset /proc/$spid/cgroup | cut -d: -f3)
> >         echo "Cgroup cpuset: $(cat /sys/fs/cgroup/cpuset${cgpath}/cpuset.cpus)"
> >     done; 
> > done
> > ```
> 
> @OhmSpectator Updated this doc:https://docs.google.com/document/d/1qoCq7SPk6UR2LCb6APVbmLLgo7W7wONJUncTrXCbPeM/edit?tab=t.0#heading=h.p1hte28jbxjx with the CPUset group info and CPU Affinity. Please refer to it. I don't see any issue with it so far.

@OhmSpectator @rene @eriknordmark Please kickoff the tests.

**Review by eriknordmark — APPROVED** [review]
Re-run tests

**Comment by OhmSpectator** [comment]
@roja-zededa, thanks for the testsing! Interestingly, we enforce pinning more by setting CPUSet in a cgroup now rather than by setting affinity in QEMU. But let it be.

**Review by OhmSpectator — APPROVED** [review]
